File: services/ai-storage-service/src/services/storageService.js
const { v4: uuid } = require('uuid');
const path = require('path');
const { getContainerClient } = require('../config/azureBlob');
const FileAsset = require('../models/FileAsset');
const env = require('../config/env');

function sanitizeFilename(name) {
  return name.replace(/[^a-zA-Z0-9._-]/g, '_').slice(-150);
}

// Streams the buffer straight to Blob Storage (uploadData here is fine for the
// multer-memory buffer we receive; for very large files switch multer to disk
// storage + client.uploadStream so nothing large sits fully in process memory).
async function uploadFile({ ownerId, buffer, originalName, mimeType }) {
  const client = getContainerClient();
  if (!client) {
    const err = new Error('Azure Blob Storage not configured');
    err.status = 503;
    throw err;
  }
  const safeName = sanitizeFilename(originalName || 'upload');
  const blobName = `${ownerId}/${uuid()}${path.extname(safeName)}`;
  const blockBlobClient = client.getBlockBlobClient(blobName);

  await blockBlobClient.uploadData(buffer, {
    blobHTTPHeaders: { blobContentType: mimeType },
  });

  const asset = await FileAsset.create({
    ownerId,
    blobName,
    originalName: safeName,
    mimeType,
    sizeBytes: buffer.length,
    container: env.azure.container,
  });

  return asset;
}

async function getDownloadStream({ ownerId, id }) {
  const asset = await FileAsset.findOne({ _id: id, ownerId });
  if (!asset) return null;
  const client = getContainerClient();
  if (!client) {
    const err = new Error('Azure Blob Storage not configured');
    err.status = 503;
    throw err;
  }
  const blockBlobClient = client.getBlockBlobClient(asset.blobName);
  const download = await blockBlobClient.download();
  return { asset, stream: download.readableStreamBody };
}

async function deleteFile({ ownerId, id }) {
  const asset = await FileAsset.findOne({ _id: id, ownerId });
  if (!asset) return false;
  const client = getContainerClient();
  if (client) {
    const blockBlobClient = client.getBlockBlobClient(asset.blobName);
    await blockBlobClient.deleteIfExists();
  }
  await asset.deleteOne();
  return true;
}

module.exports = { uploadFile, getDownloadStream, deleteFile };

File: services/ai-storage-service-python/services/storage_service.py
import uuid
import os
import re
from datetime import datetime
from config.azure_blob import get_container_client
from config.database import get_database
from config.settings import settings
from fastapi import HTTPException
from bson import ObjectId
from azure.storage.blob import ContentSettings

def sanitize_filename(name: str) -> str:
    safe_name = re.sub(r'[^a-zA-Z0-9._-]', '_', name)
    return safe_name[-150:]

async def upload_file(owner_id: str, file_bytes: bytes, original_name: str, mime_type: str) -> dict:
    client = get_container_client()
    if not client:
        raise HTTPException(status_code=503, detail="Azure Blob Storage not configured")
        
    safe_name = sanitize_filename(original_name or "upload")
    _, ext = os.path.splitext(safe_name)
    blob_name = f"{owner_id}/{uuid.uuid4()}{ext}"
    
    blob_client = client.get_blob_client(blob_name)
    
    await blob_client.upload_blob(
        file_bytes, 
        overwrite=True, 
        content_settings=ContentSettings(content_type=mime_type)
    )
    
    db = get_database()
    asset = {
        "ownerId": owner_id,
        "blobName": blob_name,
        "originalName": safe_name,
        "mimeType": mime_type,
        "sizeBytes": len(file_bytes),
        "container": settings.azure_container,
        "metadata": {},
        "createdAt": datetime.utcnow(),
        "updatedAt": datetime.utcnow()
    }
    
    result = await db.files.insert_one(asset)
    asset["_id"] = str(result.inserted_id)
    return asset

async def get_download_stream(owner_id: str, file_id: str) -> tuple:
    db = get_database()
    try:
        obj_id = ObjectId(file_id)
    except:
        return None, None
        
    asset = await db.files.find_one({"_id": obj_id, "ownerId": owner_id})
    if not asset:
        return None, None
        
    client = get_container_client()
    if not client:
        raise HTTPException(status_code=503, detail="Azure Blob Storage not configured")
        
    blob_client = client.get_blob_client(asset["blobName"])
    stream = await blob_client.download_blob()
    
    asset["_id"] = str(asset["_id"])
    return asset, stream

async def delete_file(owner_id: str, file_id: str) -> bool:
    db = get_database()
    try:
        obj_id = ObjectId(file_id)
    except:
        return False
        
    asset = await db.files.find_one({"_id": obj_id, "ownerId": owner_id})
    if not asset:
        return False
        
    client = get_container_client()
    if client:
        blob_client = client.get_blob_client(asset["blobName"])
        try:
            await blob_client.delete_blob()
        except:
            pass
            
    await db.files.delete_one({"_id": obj_id})
    return True

async def list_files(owner_id: str, limit: int = 100) -> list:
    db = get_database()
    cursor = db.files.find({"ownerId": owner_id}).sort("createdAt", -1).limit(limit)
    files = []
    async for doc in cursor:
        doc["_id"] = str(doc["_id"])
        files.append(doc)
    return files
